Unable to exec commands for dependencies of a sub-package in a monorepo

<!--
  If this issue affects many people in a company/big team, create a post for your company in the following discussion:
  https://github.com/pnpm/pnpm/discussions/3787
  and link the issue in your post.

  This will help us prioritize issues that affect more people.
-->

### pnpm version: 6.32.3

### Code to reproduce the issue:

<!--
If there was a fatal error also include a gist of your node_modules/.pnpm-debug.log file.
-->

For the examples below, `del-cli` is a dependency in the root `package.json` file. `msw` is only a dependency in the `example` `package.json` file.

```bash
$ pnpm --filter example exec msw init
 ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL  not found: msw
```

```bash
$ pnpm --filter example exec del-cli "node_modules"
```

### Expected behavior:

If using `--filter`, it should be possible to execute commands with dependencies only in those filtered packages and not in the root `package.json`.

### Actual behavior:

The `msw` command fails while the `del-cli` command succeeds. It seems like only dependencies in the root `package.json` works with `exec`.

### Additional information:

 - `node -v` prints: v16.13.2
 - Windows, macOS, or Linux?: macOS

--- a/.changeset/strange-llamas-tie.md
+++ b/.changeset/strange-llamas-tie.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/plugin-commands-script-runners": patch
+"pnpm": patch
+---
+
+When `pnpm exec` is running a command in a workspace project, the commands that are in the dependencies of that workspace project should be in the PATH [#4481](https://github.com/pnpm/pnpm/issues/4481).

--- a/packages/plugin-commands-script-runners/src/exec.ts
+++ b/packages/plugin-commands-script-runners/src/exec.ts
@@ -137,7 +137,7 @@ export async function handler (
               PNPM_PACKAGE_NAME: opts.selectedProjectsGraph?.[prefix]?.package.manifest.name,
             },
             prependPaths: [
-              path.join(opts.dir, 'node_modules/.bin'),
+              path.join(prefix, 'node_modules/.bin'),
               ...opts.extraBinPaths,
             ],
             userAgent: opts.userAgent,

--- a/packages/plugin-commands-script-runners/test/exec.e2e.ts
+++ b/packages/plugin-commands-script-runners/test/exec.e2e.ts
@@ -52,7 +52,7 @@ test('pnpm recursive exec', async () => {
   ])
 
   const { selectedProjectsGraph } = await readProjects(process.cwd(), [])
-  await execa('pnpm', [
+  await execa(pnpmBin, [
     'install',
     '-r',
     '--registry',
@@ -72,6 +72,45 @@ test('pnpm recursive exec', async () => {
 
   expect(outputs1).toStrictEqual(['project-1', 'project-2-prebuild', 'project-2', 'project-2-postbuild'])
   expect(outputs2).toStrictEqual(['project-1', 'project-3'])
+})
+
+test('pnpm recursive exec finds bin files of workspace projects', async () => {
+  preparePackages([
+    {
+      name: 'project-1',
+      version: '1.0.0',
+
+      dependencies: {
+        cowsay: '1.5.0',
+      },
+    },
+    {
+      name: 'project-2',
+      version: '1.0.0',
+
+      dependencies: {
+        cowsay: '1.5.0',
+      },
+    },
+  ])
+
+  const { selectedProjectsGraph } = await readProjects(process.cwd(), [])
+  await execa(pnpmBin, [
+    'install',
+    '-r',
+    '--registry',
+    REGISTRY,
+    '--store-dir',
+    path.resolve(DEFAULT_OPTS.storeDir),
+  ])
+  await exec.handler({
+    ...DEFAULT_OPTS,
+    dir: process.cwd(),
+    recursive: true,
+    selectedProjectsGraph,
+  }, ['cowsay', 'hi'])
+
+  // If there was no exception, the test passed
 })
 
 test('exec inside a workspace package', async () => {
@@ -115,7 +154,7 @@ test('exec inside a workspace package', async () => {
     },
   ])
 
-  await execa('pnpm', [
+  await execa(pnpmBin, [
     'install',
     '-r',
     '--registry',
@@ -196,7 +235,7 @@ test('testing the bail config with "pnpm recursive exec"', async () => {
   ])
 
   const { selectedProjectsGraph } = await readProjects(process.cwd(), [])
-  await execa('pnpm', [
+  await execa(pnpmBin, [
     'install',
     '-r',
     '--registry',
@@ -272,7 +311,7 @@ test('pnpm recursive exec --no-sort', async () => {
   ])
 
   const { selectedProjectsGraph } = await readProjects(process.cwd(), [])
-  await execa('pnpm', [
+  await execa(pnpmBin, [
     'install',
     '-r',
     '--registry',

